llvm/include/llvm/Analysis/DXILResource.h

```diff
@@ -574,6 +577,81 @@ class DXILResourceBindingWrapperPass : public ModulePass {
 
 ModulePass *createDXILResourceBindingWrapperPassPass();
 
+enum class ResourceCounterDirection {
+  Increment,
+  Decrement,
+  Unknown,
+  Invalid,
+};
+
+class DXILResourceCounterDirectionMap {
+  std::vector<std::pair<dxil::ResourceBindingInfo, ResourceCounterDirection>>
+      CounterDirections;
+
+public:
+  void populate(Module &M, DXILBindingMap &DBM);
+
+  ResourceCounterDirection
+  operator[](const dxil::ResourceBindingInfo &Info) const {
+    auto Lower = std::lower_bound(
+        CounterDirections.begin(), CounterDirections.end(),
+        std::pair{Info, ResourceCounterDirection::Unknown},
+        [](auto LHS, auto RHS) { return LHS.first < RHS.first; });
```

Comment:
You can simplify this in a couple of ways:
1. `llvm::lower_bound` accepts a range argument
2. `std::lower_bound` does not require that the type of the value to compare against matches the type of the collection (See the examples on [cppreference](https://en.cppreference.com/w/cpp/algorithm/lower_bound))

Also, it's better to use `const auto &` here rather than `auto` to avoid copies.

```suggestion
    auto Lower = llvm::lower_bound(
        CounterDirections, Info,
        [](const auto &LHS, const auto &RHS) { return LHS.first < RHS; });
```